parlai/core/torch_ranker_agent.py

```diff
@@ -0,0 +1,351 @@
+            self.build_model()
+            if model_file:
+                print('Loading existing model parameters from ' + model_file)
+                self.load(model_file)
+
+        self.rank_loss = nn.CrossEntropyLoss(reduction='sum')
+        self._set_fixed_candidates(shared)
+
+        if self.use_cuda:
+            self.model.cuda()
+            self.rank_loss.cuda()
+
+        optim_params = [p for p in self.model.parameters() if p.requires_grad]
+        self._init_optim(optim_params)
+
+    def score_candidates(self, batch, cand_vecs):
+        """Given a batch and candidate set, return scores (for ranking)"""
+        raise NotImplementedError(
+            'Abstract class: user must implement score()')
+
+    def build_model(self):
+        """Build a new model (implemented by children classes)"""
+        raise NotImplementedError(
+            'Abstract class: user must implement build_model()')
+
+    def train_step(self, batch):
+        """Train on a single batch of examples."""
+        if batch.text_vec is None:
+            return
+        batchsize = batch.text_vec.size(0)
+        self.model.train()
+        self.optimizer.zero_grad()
+
+        cands, cand_vecs, label_inds = self._build_candidates(batch,
+            source=self.opt['train_candidates'], mode='train')
+        scores = self.score_candidates(batch, cand_vecs)
+        loss = self.rank_loss(scores, label_inds)
+
+        self.metrics['loss'] += loss.item()
+        self.metrics['examples'] += batchsize
+        _, ranks = scores.sort(1, descending=True)
+        for b in range(batchsize):
+            rank = (ranks[b] == label_inds[b]).nonzero().item()
+            self.metrics['rank'] += 1 + rank
+
+        loss.backward()
+        self.update_params()
+
+        # get predictions but not full rankings--too slow to get hits@1 score
+        preds = [cands[row[0]] for row in ranks]
+
+        return Output(preds)
+
+    @torch.no_grad()
+    def eval_step(self, batch):
+        """Evaluate a single batch of examples."""
+        if batch.text_vec is None:
+            return
+        batchsize = batch.text_vec.size(0)
+        self.model.eval()
+
+        cands, cand_vecs, label_inds = self._build_candidates(batch,
+            source=self.opt['eval_candidates'], mode='eval')
+
+        scores = self.score_candidates(batch, cand_vecs)
+        _, ranks = scores.sort(1, descending=True)
+
+        if label_inds is not None:
+            loss = self.rank_loss(scores, label_inds)
+            self.metrics['loss'] += loss.item()
+            self.metrics['examples'] += batchsize
+            for b in range(batchsize):
+                rank = (ranks[b] == label_inds[b]).nonzero().item()
+                self.metrics['rank'] += 1 + rank
+
+        cand_preds = [[cands[i] for i in row] for row in ranks]
+        preds = [row[0] for row in cand_preds]
+        return Output(preds, cand_preds)
+
+    def _build_candidates(self, batch, source, mode):
```

Comment:
very nice. We might need to check that this docstring is RST friendly, but i like it a lot

Reply:
just marking resolved